testsuite/tests/lib-smallint/test_int8.ml

```diff
@@ -0,0 +1,143 @@
+(* TEST include stdlib_beta; flags = "-extension small_numbers_beta"; *)
+
+module Int8 = Stdlib_beta.Int8
+
+let int_size = Int8.int_size
+
+let () = assert (0 < int_size && int_size < Sys.int_size)
+
+let max_int = (1 lsl (int_size - 1)) - 1
+
+let min_int = lnot max_int
+
+let mask = (1 lsl int_size) - 1
+
+let to_int x : int =
+  let i : int = Int8.to_int x in
+  assert (Obj.repr i == Obj.repr x);
+  assert (min_int <= i && i <= max_int);
+  i
+
+let of_int (i : int) =
+  let x = Int8.of_int i in
+  assert (to_int x land mask == i land mask);
+  x
+
+let rng = Random.State.make [| int_size |]
+
+(** sparse test cases, concentrated around 0 and the endpoints *)
+let test_cases =
+  let is_even = 1 - (int_size land 1) in
+  List.init (int_size - is_even) (fun size ->
+      let bit = 1 lsl size in
+      let rand () =
+        Random.State.int_in_range rng ~min:(bit lsr 1) ~max:(bit - 1)
+      in
+      [rand (); lnot (rand ()); max_int - rand (); lnot (max_int - rand ())])
+  |> List.concat |> List.sort Int.compare
+
+let test1 f = ListLabels.iter test_cases ~f
+
+let test2 f = test1 (fun x -> test1 (fun y -> f x y))
+
+let test_round_trip () =
+  let test hi lo =
+    let hi = hi lsl int_size in
+    assert (lo == to_int (of_int (hi lxor lo)))
+  in
+  test2 (fun hi lo ->
+      (* generate test cases with different hi bits *)
+      test hi lo;
+      test (Random.bits ()) lo)
+
+let equal_arith x i = x == of_int i
+
+let equal_logical x i = to_int x == i
+
+let same_float x y = Int64.equal (Int64.bits_of_float x) (Int64.bits_of_float y)
+
+let assert_equal equal x y =
+  let x = try Ok (x ()) with exn -> Error exn in
+  let y = try Ok (y ()) with exn -> Error exn in
+  match x, y with
+  | Ok x, Ok y -> assert (equal x y)
+  | Error exn, Error exn' -> assert (exn = exn')
+  | Ok _, Error exn | Error exn, Ok _ -> raise exn
+
+let test_conv1 int8_f int_f ~equal =
+  test1 (fun x ->
+      assert_equal equal (fun () -> int8_f (of_int x)) (fun () -> int_f x))
+
+let test_conv2 int8_f int_f ~equal =
+  test2 (fun x y ->
+      assert_equal equal
+        (fun () -> int8_f (of_int x) (of_int y))
+        (fun () -> int_f x y))
+
+let test_arith1 = test_conv1 ~equal:equal_arith
+
+let test_arith2 = test_conv2 ~equal:equal_arith
+
+let test_logical1 = test_conv1 ~equal:equal_logical
+
+let test_logical2 = test_conv2 ~equal:equal_logical
+
+let reference_shift_right_logical x i =
+  (* we need to ensure that we shift in zero bytes, which is incompatible with
+     sign-extension *)
+  Int.shift_right_logical (if i > 0 then x land mask else x) i
+
+let () =
```

Comment:
Things defined in the library that aren't tested here: `neg`, `min_int`, and `max_int`.

Testing `neg` seems straightforward.  For `min_int` and `max_int` I guess you can just compare them to the concrete expected values.